lib/excheck/io_server.ex

```diff
@@ -0,0 +1,148 @@
+defmodule ExCheck.IOServer do
+  use GenServer
+  alias __MODULE__, as: State
+  alias ExCheck.ErrorAgent, as: ErrAgent
+
+  @moduledoc """
+  Process which manipulates IO output and forwards it to previous group leader.
+  """
+
+  defstruct gl: :no_pid,     # Previous group leader process
+            pids: [],        # The pids from which we are redirecting output
+            tests: 0,        # Amount of triq property tests ran
+            enabled: false,  # Colors enabled in terminal?
+            agent: :no_pid
+
+  @server __MODULE__
+
+  # API
+
+  @doc """
+  Starts a new server process, links it to the calling process.
+  """
+  def start_link do
+    GenServer.start_link(__MODULE__, :ok, name: @server)
+  end
+
+  @doc "Redirects IO output from pid to this server."
+  def redirect(pid) do
+    @server |> GenServer.call {:redirect, pid}
+  end
+
+  @doc "Get the total amount of property tests that ran so far."
+  def total_tests do
+    @server |> GenServer.call :total_tests
+  end
+
+  @doc "Returns all error loggings from property tests"
+  def errors do
+    @server |> GenServer.call :errors
+  end
+
+  # Callbacks
+
+  @doc false
+  def init(:ok) do
+    {:ok, agent} = ErrAgent.start_link
+    gl = Process.group_leader        # get previous group leader
+    {:ok, %State{gl: gl, enabled: IO.ANSI.enabled?, agent: agent}}
+  end
+
+  @doc false
+  def terminate(_reason, %State{gl: gl, pids: pids, agent: agent}) do
+    for pid <- pids do
+      Process.group_leader(pid, gl)  # Revert back to old group leader
+    end
+    ErrAgent.stop(agent)
+    :ok
+  end
+
+  @doc false
+  def handle_call({:redirect, pid}, _from, state = %State{pids: pids}) do
+    Process.group_leader(pid, self)  # redirect all IO from pid to this process
+    {:reply, :ok, %State{state | pids: [pid | pids]}}
+  end
+  def handle_call(:total_tests, _from, state = %State{tests: tests}) do
+    {:reply, tests, state}
+  end
+  def handle_call(:errors, _from, state = %State{agent: agent}) do
+    response = ErrAgent.errors(agent)
+    {:reply, response, state}
+  end
+
+  @doc false
+  def handle_info({:io_request, from, ref, 
+                  {:put_chars, :unicode, :io_lib, :format, [msg, value_list]}}, state) do
+    # Receive all IO requests here
+    new_state = handle_output(msg, value_list, from, state)
+    from |> send {:io_reply, ref, :ok}
+    {:noreply, new_state}
+  end
+  def handle_info({:io_request, from, ref, 
+                  {:put_chars, :unicode, msg}}, state) do
+    # Receive all IO requests here
+    new_state = handle_output(msg, [], from, state)
+    from |> send {:io_reply, ref, :ok}
+    {:noreply, new_state}
+  end
+  # TODO need more cases here to handle all forms of IO?
```

Comment:
The server itself doesn't expect any other kind of messages besides :io_request.

Top clause handles all output coming from triq (using io:format). Bottom clause handles Elixir output (for example using IO.write). The differences are that io:format uses a list of integers with interpolated values in a separate value list while Elixir uses (utf8) binaries that already contain the interpolated data. (e.g. "#{12345}"). It looks like Elixir only uses put_chars..

Should we support the input requests? (Seems rather useless in a test run but you never know..)
Not sure if this covers everything, for now I will add a fallback clause just to be safe (won't handle input requests). 

For a more complete solution, we might have to mimic something like this: https://github.com/erlang/otp/blob/1523be48ab4071b158412f4b06fe9c8d6ba3e73c/lib/kernel/src/user.erl#L182-#L291